src/Service/Bulk/BulkCreateUsersService.php

```diff
@@ -0,0 +1,294 @@
+    }
+
+    /**
+     * @throws LineItemNotFoundException
+     */
+    public function createUsers(
+        array $lineItemIds,
+        array $lineItemSlugs,
+        array $userPrefixes,
+        int $batchSize,
+        ?string $groupPrefix
+    ): UserCreationResult {
+
+        $notExistLineItemsArray = $userGroupIds = [];
+
+        if (!empty($lineItemIds) || !empty($lineItemSlugs)) {
+            $this->lineItemIds = $lineItemIds;
+            $this->lineItemSlugs = $lineItemSlugs;
+            $criteria = $this->getFindLineItemCriteria();
+            $lineItems = $this->lineItemRepository->findLineItemsByCriteria($criteria)->jsonSerialize();
+            if (empty($lineItems)) {
+                $exceptionMessage = $this->lineItemIds
+                    ? implode(',', $this->lineItemIds) . ' Line item id(s) not exist in the system'
+                    : implode(',', $this->lineItemSlugs) . ' Line item slug(s) not exist in the system';
+                throw new LineItemNotFoundException($exceptionMessage);
+            }
+
+            $this->generateSlugData($lineItems);
+            $notExistLineItemsArray = $this->lineItemIds
+                ? array_diff($lineItemIds, array_keys($this->lineItemSlugs))
+                : array_diff($lineItemSlugs, array_values($this->lineItemSlugs));
+        }
+
+        if (empty($lineItemIds) && empty($lineItemSlugs)) {
+            $this->getAllLineItemSlugs();
+        }
+
+        $userNameLastIndexes = $this->getLastUserAssignedToLineItems($this->lineItemSlugs);
+
+        $userGroupAssignCount = 0;
+        if ($groupPrefix) {
+            $loadBalancer = $this->ltiInstanceLoadBalancerFactory->__invoke('userGroupId');
+            $userGroupIds = $loadBalancer->generateGroupIds($groupPrefix);
+            $userGroupAssignCount = (int) ceil(
+                count($userPrefixes) * count($this->lineItemSlugs) * $batchSize / count($userGroupIds)
+            );
+        }
+
+        $slugTotalUsers = $this->createBulkUserAssignmentData(
+            $userNameLastIndexes,
+            $userGroupIds,
+            $userPrefixes,
+            $userGroupAssignCount,
+            $groupPrefix,
+            $batchSize
+        );
+
+        $message = $this->getOperationResultMessage($slugTotalUsers, $userPrefixes);
+
+        return new UserCreationResult($message, $notExistLineItemsArray);
+    }
+
+    public function getOperationResultMessage(array $slugTotalUsers, array $userPrefix): string
+    {
+        $message = '';
+        $userPrefixString = implode(',', $userPrefix);
+
+        foreach ($slugTotalUsers as $slug => $totalUsers) {
+            $message .= sprintf(
+                "%s users created for line item %s for user prefix %s \n",
+                $totalUsers,
+                $slug,
+                $userPrefixString
+            );
+        }
+
+        return $message;
+    }
+
+
+    private function generateSlugData(array $lineItems): void
```

Comment:
I wouldn't say that the method generates something

Reply:
It is generating a line item slug data array and assigning it to a class variable $this->lineItemSlugs . So it doesn't returns anything